File: swift/tree/insertAndRemove.swift
class TreeNode {
  var val: Int
  var left: TreeNode?
  var right: TreeNode?

  init(_ val: Int) {
    self.val = val
    left = nil
    right = nil
  }
}

// Insert a new node and return the root of the BST
func insert(_ root: TreeNode?, _ val: Int) -> TreeNode? {
  if root == nil {
    return TreeNode(val)
  }

  if val > root!.val {
    root?.right = insert(root?.right, val)
  }
  else if val < root!.val {
    root?.left = insert(root?.left, val)
  }
  return root
}

// Return the minimum value node of the BST
func minValueNode(_ root: TreeNode?) -> TreeNode? {
  var curr = root
  while curr != nil && curr?.left != nil {
    curr = curr?.left
  }
  return curr
}

// Remove a node and return the root of the BST
func remove(_ root: TreeNode?, _ val: Int) -> TreeNode? {
  if root == nil {
    return TreeNode(val)
  }

  if val > root!.val {
    root?.right = remove(root?.right, val)
  }
  else if val < root!.val {
    root?.left = remove(root?.left, val)
  }
  else {
    if root?.left == nil {
      return root?.right
    }
    else if root?.right == nil {
      return root?.left
    }
    else {
      let minNode = minValueNode(root?.right)
      root?.val = minNode!.val
      root?.right = remove(root?.right, minNode!.val)
    }
  }
  return root
}

File: csharp/tree/InsertAndRemove.cs


public class InsertAndRemove
{
    // Insert a new node and return the root of the BST.
    public TreeNode Insert(TreeNode root, int val)
    {
        if (root == null)
        {
            return new TreeNode(val);
        }

        if (val > root.Value)
        {
            root.Right = Insert(root.Right, val);
        }
        else if (val < root.Value)
        {
            root.Left = Insert(root.Left, val);
        }
        return root;
    }

    // Return the minimum value node of the BST.
    public TreeNode MinValueNode(TreeNode root)
    {
        TreeNode curr = root;
        while (curr != null && curr.Left != null)
        {
            curr = curr.Left;
        }
        return curr;
    }

    // Remove a node and return the root of the BST.
    public TreeNode Remove(TreeNode root, int val)
    {
        if (root == null)
        {
            return null;
        }
        if (val > root.Value)
        {
            root.Right = Remove(root.Right, val);
        }
        else if (val < root.Value)
        {
            root.Left = Remove(root.Left, val);
        }
        else
        {
            if (root.Left == null)
            {
                return root.Right;
            }
            else if (root.Right == null)
            {
                return root.Left;
            }
            else
            {
                TreeNode minNode = MinValueNode(root.Right);
                root.Value = minNode.Value; 
                root.Right = Remove(root.Right, minNode.Value);
            }
        }
        return root;
    }
}